### PR TITLE
PMM-5757 diagnostic data collector

### PR DESCRIPTION
**Pull request by percona-csalguero**
https://jira.percona.com/browse/PMM-5757

- Diagnostic data collector
- Replicaset status collector
- Collstats collector
- Server status collector
- Main parameters
- HTTP server that will support having multiple paths to expose metrics

### CODE REVIEW

File: exporter/exporter.go [hunk 1]
```diff
@@ -13,36 +18,74 @@ const (
...
+	exp.collectors = append(exp.collectors, &diagnosticDataCollector{ctx: ctx, client: client})
+	exp.collectors = append(exp.collectors, &replSetGetStatusCollector{ctx: ctx, client: client})
+	if len(opts.CollStatsCollections) > 0 {
+		exp.collectors = append(exp.collectors, &collstatsCollector{ctx: ctx, client: client, collections: opts.CollStatsCollections})
+	}
+
+	return exp, nil
+}
+
+func (e *Exporter) Run() {
```
Comment (AlekSi) [inline exporter/exporter.go]: That really should be in #4

**Review by BupycHuk — CHANGES_REQUESTED** [review]
Please fix merge conflicts

File: exporter/collstats_collector.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	for _, dbCollection := range d.collections {
+		parts := strings.Split(dbCollection, ".")
+		if len(parts) != 2 {
+			continue
+		}
+
+		cmd := bson.D{{Key: "collStats", Value: parts[1]}}
+		res := d.client.Database(parts[0]).RunCommand(d.ctx, cmd)
+		var m bson.M
+		if err := res.Decode(&m); err != nil {
```
[the remark this answers is not in the document]
Comment (BupycHuk) [reply exporter/collstats_collector.go:49]: maybe it's better to disable this linter

File: exporter/replset_status_collector_test.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+	"github.com/prometheus/client_golang/prometheus/testutil"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestReplsetStatusCollector(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	client := getTestClient(ctx, t)
```
Comment (BupycHuk) [inline exporter/replset_status_collector_test.go]: Can we add test for ReplsetStatusCollector for cases with no replicaset?

File: exporter/serverstatus_collector_test.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/stretchr/testify/assert"
+)
+
+func TestServerStatusDataCollector(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	client := getTestClient(ctx, t)
+
+	c := &diagnosticDataCollector{
```
Comment (BupycHuk) [inline exporter/serverstatus_collector_test.go]: `diagnosticDataCollector`?

**Review by BupycHuk — CHANGES_REQUESTED** [review]
I remember that you said that you will add API as part of this PR, but I don't see it. Will you make separate PR?

File: internal/tu/testutils.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
+package tu
```
Comment (BupycHuk) [inline internal/tu/testutils.go:18]: what does tu means?

[the remark this answers is not in the document]
Comment (BupycHuk) [reply internal/tu/testutils.go:18]: please fix golangci linter comments or add ignore comments.

File: exporter/replset_status_collector_test.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	client := tu.TestClient(ctx, tu.MongoDBStandAlonePort, t)
+
+	c := &replSetGetStatusCollector{
+		ctx:    ctx,
+		client: client,
+	}
+
+	expected := strings.NewReader(``)
+	err := testutil.CollectAndCompare(c, expected)
+	assert.Error(t, err)
```
Comment (BupycHuk) [inline exporter/replset_status_collector_test.go]: let's use `assert.EqualError`

**Review by BupycHuk — CHANGES_REQUESTED** [review]
I don't see anything related to `HTTP server that will support having multiple paths to expose metrics`

File: internal/tu/testutils.go (hunk 5 above)
Comment (percona-csalguero) [reply internal/tu/testutils.go:18]: test utils

File: exporter/replset_status_collector_test.go (hunk 6 above)
Comment (percona-csalguero) [reply exporter/replset_status_collector_test.go]: Improved to detect the Replicaset not yet initialized error

**Review by BupycHuk — APPROVED** [review]

**Comment by BupycHuk** [comment]
please fix a build.

File: exporter/collstats_collector.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply exporter/collstats_collector.go:49]: Agree

@percona-csalguero ^

**Review by AlekSi — CHANGES_REQUESTED** [review]
CI should pass

**Comment by percona-csalguero** [comment]
@AlekSi @BupycHuk , CI checks passing

File: exporter/collstats_collector.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlekSi) [reply exporter/collstats_collector.go:49]: @percona-csalguero ^^^

**Review by AlekSi — CHANGES_REQUESTED** [review]

**Review by BupycHuk — APPROVED** [review]

**Review by AlekSi — APPROVED** [review]

**Review by askomorokhov — APPROVED** [review]
👍

**Pull request merged**